text/3470-crabi-v1.md

```diff
@@ -0,0 +1,661 @@
+functionality.
+
+# C-based ABI
+[c-based-abi]: #c-based-abi
+
+crABI uses the target's C ABI as a base. For example, on x86-64 Linux, crABI
+uses the [x86-64 psABI](https://gitlab.com/x86-psABIs/x86-64-ABI) as a base,
+just like `extern "C"` does.
+
+Any type whose ABI is already defined by C will be passed through crABI
+identically. Types defined by crABI that the C ABI does not support will be
+translated into a representation using types the C ABI supports (potentially
+indirectly via other crABI-supported types).
+
+crABI does not extend the target C ABI in any ways that would require the use
+of target-specific assembly language to support, such as the use of CPU flags
+or registers not already defined by the C ABI. Doing so would require specific
+enablement work in other languages or runtimes, rather than just building on
+existing support for C FFI.
+
+# Rust language interface and interaction with other features
+[language-interface]: #language-interface
+
+The Rust implementation of crABI will include:
+- An ABI for defining functions or referencing external functions:
+  `extern "crabi"`.
+- A repr for laying out data structures (`struct`, `union`, `enum`) compatible
+  with crABI: `repr(crabi)`.
+- A new lint `improper_crabi_types`, analogous to `improper_ctypes`. This lint
+  defaults to `deny`.
+
+crABI structures and functions also support any structure defined using
+`repr(C)`, as well as enums that have a discriminant type specified.
+
+crABI structures and functions support types defined using `repr(transparent)`
+if they would support the type of the field.
+
+For an enum, the crABI `repr` allows additionally specifying the discriminant
+type: `repr(crabi, u8)`. If an `enum` specifies `repr(crabi)` but does not
+specify a discriminant type, the enum is guaranteed to use the smallest
+discriminant type that holds the maximum discriminant value used by a variant
+in the enum. (This differs from the behavior of `repr(C)` enums without a
+discriminant type.)
+
+crABI supports passing and storing pointers to `repr(Rust)` structures, to
+allow using these opaque pointers as handles, but does not support passing such
+structures by value.
+
+crABI works in combination with `repr(align)`, with the same meaning.
+
+crABI does not support `repr(packed)`.
+
+# crABI versioning and evolution
+[crabi-versioning-and-evolution]: #crabi-versioning-and-evolution
+
+crABI has has a major and minor version number, similar to semver.
+
+The major version number will change if crABI changes *incompatibly*; for
+instance, crABI 2.0 would not necessarily interoperate with crABI 1.3.
+
+The minor version number will change if crABI changes *compatibly*, such that
+newer callers can understand and work with older functions and structures.
+Older callers cannot necessarily understand and work with newer functions and
+structures, unless the newer side restricts itself to features understood by an
+older version.
+
+This RFC defines crABI 1.0.
```

Comment:
I know that we feel pretty confident in most of the decisions in this RFC, but since crABI has not seen actual usage, especially regarding interop with other languages, it feels prudent to me if we started at `0.1.0` and allowed ourselves breaking changes, and only decided to move to `1.0` once enough (different) eyes have been laid on the problem.